clippy_lints/src/eval_order_dependence.rs

```diff
@@ -56,6 +77,75 @@ impl LateLintPass for EvalOrderDependence {
             _ => {}
         }
     }
+    fn check_stmt(&mut self, cx: &LateContext, stmt: &Stmt) {
+        match stmt.node {
+            StmtExpr(ref e, _) | StmtSemi(ref e, _) => DivergenceVisitor(cx).maybe_walk_expr(e),
+            StmtDecl(ref d, _) => {
+                if let DeclLocal(ref local) = d.node {
+                    if let Local { init: Some(ref e), .. } = **local {
+                        DivergenceVisitor(cx).visit_expr(e);
+                    }
+                }
+            },
+        }
+    }
+}
+
+struct DivergenceVisitor<'a, 'tcx: 'a>(&'a LateContext<'a, 'tcx>);
+
+impl<'a, 'tcx> DivergenceVisitor<'a, 'tcx> {
+    fn maybe_walk_expr(&mut self, e: &Expr) {
+        match e.node {
+            ExprClosure(..) => {},
+            ExprMatch(ref e, ref arms, _) => {
+                self.visit_expr(e);
+                for arm in arms {
+                    if let Some(ref guard) = arm.guard {
+                        self.visit_expr(guard);
+                    }
+                    // make sure top level arm expressions aren't linted
+                    walk_expr(self, &*arm.body);
+                }
+            }
+            _ => walk_expr(self, e),
+        }
+    }
+    fn report_diverging_sub_expr(&mut self, e: &Expr) {
+        span_lint(
+            self.0,
+            DIVERGING_SUB_EXPRESSION,
+            e.span,
+            "sub-expression diverges",
+        );
+    }
+}
+
+impl<'a, 'tcx, 'v> Visitor<'v> for DivergenceVisitor<'a, 'tcx> {
+    fn visit_expr(&mut self, e: &'v Expr) {
+        // this match can be replaced by just the default arm, once
+        // https://github.com/rust-lang/rust/issues/35121 makes sure that
+        // ! is propagated properly
+        match e.node {
+            ExprAgain(_) |
+            ExprBreak(_) |
+            ExprRet(_) => self.report_diverging_sub_expr(e),
+            ExprCall(ref func, _) => match self.0.tcx.expr_ty(func).sty {
+                ty::TyFnDef(_, _, fn_ty) |
+                ty::TyFnPtr(fn_ty) => if let ty::TyNever = self.0.tcx.erase_late_bound_regions(&fn_ty.sig).output.sty {
+                    self.report_diverging_sub_expr(e);
+                },
+                _ => {},
+            },
+            ExprMethodCall(..) => { /* TODO */ },
```

Comment:
No... It's hard to do, because we don't have a method handle, just the name and the type to dispatch on